x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -732,40 +786,71 @@ private void innerRefresh(String tokenDocId, Authentication userAuth, ActionList
-                                            Throwable cause = ExceptionsHelper.unwrapCause(e);
-                                            if (cause instanceof VersionConflictEngineException ||
-                                                isShardNotAvailableException(e)) {
-                                                innerRefresh(tokenDocId, userAuth,
-                                                    listener, attemptCount);
-                                            } else {
-                                                onFailure.accept(e);
-                                            }
-                                        }),
-                                    client::update);
+                            if (eligibleForMultiRefresh(source)) {
+                                final Map<String, Object> refreshTokenSrc = (Map<String, Object>) source.get("refresh_token");
+                                final String supersedingTokenDocId = (String) refreshTokenSrc.get("superseded_by");
+                                logger.debug("Token document [{}] was recently refreshed, attempting to reuse [{}] for returning an " +
+                                    "access token and refresh token", tokenDocId, supersedingTokenDocId);
+                                GetRequest supersedingTokenGetRequest =
+                                    client.prepareGet(SecurityIndexManager.SECURITY_INDEX_NAME, TYPE, supersedingTokenDocId).request();
+                                executeAsyncWithOrigin(client.threadPool().getThreadContext(), SECURITY_ORIGIN, supersedingTokenGetRequest,
+                                    ActionListener.<GetResponse>wrap(supersedingTokenResponse -> {
+                                        if (supersedingTokenResponse.isExists()) {
+                                            final Map<String, Object> supersedingTokenSource = supersedingTokenResponse.getSource();
+                                            final Map<String, Object> supersedingUserTokenSource = (Map<String, Object>)
+                                                ((Map<String, Object>) supersedingTokenSource.get("access_token")).get("user_token");
+                                            final Map<String, Object> supersedingRefreshTokenSrc =
+                                                (Map<String, Object>) supersedingTokenSource.get("refresh_token");
+                                            final String supersedingRefreshTokenValue = (String) supersedingRefreshTokenSrc.get("token");
+                                            reIssueTokens(supersedingUserTokenSource, supersedingRefreshTokenValue, listener);
+                                        } else {
+                                            logger.info("could not find token document [{}] for refresh", supersedingTokenDocId);
+                                            onFailure.accept(invalidGrantException("could not refresh the requested token"));
```

Comment:
you probably want to retry here otherwise we fail requests that come in between the update of the existing document and the creation of the superseding document. Additionally the "hammer" retry approach isn't good enough since it can take time for the document to get indexed.